Unify --wait strategy inputs, clarify defaults and docs; avoid SDK timeout footgun

benoittgt opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:

This PR improves the `--wait` flag by allowing explicit strategy selection and preventing SDK timeout errors.

### Changes

1. **Allow explicit `--wait=hookOnly`** - Previously not possible via CLI
2. **Add comprehensive Go documentation** for all wait strategies  
3. **Set 30s default timeout** for StatusWatcher to prevent SDK errors when timeout is not specified

### Background

Problem 1: SDK Timeout Error

When using StatusWatcher strategy without setting a timeout:
```go
upgradeClient := action.NewUpgrade(actionConfig)
upgradeClient.WaitStrategy = kube.StatusWatcherStrategy
// Fails with: "reporter failed to start: event funnel closed: context deadline exceeded"
```

Problem 2: Confusing CLI Mapping

Users couldn't explicitly set `hookOnly` and the strategy mapping was unclear ([Slack discussion](https://kubernetes.slack.com/archives/C51E88VDG/p1761307867307409)).

CLI Behavior


|CLI Option | Internal Strategy|
-- | --
--wait=watcher | watcher
--wait=legacy | legacy
--wait=hookOnly | hookOnly ⚡ now possible
No --wait flag | hookOnly (unchanged for backward compatibility)
--wait (no value) | watcher
--wait=false | hookOnly (deprecated)
--wait=true | watcher (deprecated)

</body></html>

Related
- Initial watcher strategy PR: #13604
- Related discussion: #31411

**Special notes for your reviewer**:

- Default behavior unchanged when --wait is omitted (remains hookOnly)
- SDK users can now omit timeout when using StatusWatcher (30s default applied)
- Still I think the current `--wait` VS no `--wait` can be quite confusing

**If applicable**:
- [ ] this PR contains user facing changes (the `docs needed` label should be applied if so)
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

benoittgt committed:
While testing SDK features for v4. I was surprised with the error:

> "reporter failed to start: event funnel closed: context deadline  exceeded"

It happens when you forget to set a minimal timeout:
```go
upgradeClient := action.NewUpgrade(actionConfig)
upgradeClient.WaitStrategy = kube.StatusWatcherStrategy
// When Timeout is zero, the status wait uses a context with zero timeout which
// immediately expires causing "context deadline exceeded" errors.
upgradeClient.Timeout = 2 * time.Minute
```

Also maybe it might be worth documenting this more clearly. Initial [PR](https://github.com/helm/helm/pull/13604) say:
> I have not written any docs, I assume that can be done when we are closer to Helm 4, a lot of it is covered by linking to - https://github.com/kubernetes-sigs/cli-utils/blob/master/pkg/kstatus/README.md

Related:
  - https://github.com/helm/helm/pull/31411#issuecomment-3443925663
Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Increase documentation of --wait flag

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Update pkg/cmd/flags.go

Co-authored-by: George Jenkins <gvjenkins@gmail.com>
Signed-off-by: Benoit Tigeot <benoittgt@users.noreply.github.com>
Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

TerryHowe commented on pkg/cmd/flags.go:
```diff
@@ -59,7 +59,7 @@ func AddWaitFlag(cmd *cobra.Command, wait *kube.WaitStrategy) {
 		newWaitValue(kube.HookOnlyStrategy, wait),
 		"wait",
-		"if specified, will wait until all resources are in the expected state before marking the operation as successful. It will wait for as long as --timeout. Valid inputs are 'watcher' and 'legacy'",
+		"wait until resources are ready (up to --timeout). Values: 'watcher' (default), 'legacy', and 'hookOnly'",
```
Is the default "hookOnly"?

benoittgt replied:
I think it's quite complicated to understand it quickly.


| CLI Option              | Internal Wait Strategy |
|--------------------------|------------------------|
| `--wait=watcher`         | `watcher`              | 
| `--wait=legacy`          | `legacy`               | 
| `--wait=hookOnly`        | *Not possible*         | 
| *No `--wait` flag set*   | `hookOnly`             | 
| `--wait` (no value)      | `watcher`              | 
| `--wait=false`           | `hookOnly`             | 
| `--wait=true`            | `watcher`              | 

🤯

benoittgt committed:
Make wait strategy selection more obvious

It is not clear enough that you need to set --wait=false for a hook only
strategy.

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
The default is not HookOnlyStrategy but WaitStrategy

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt commented on pkg/cmd/flags.go:
```diff
@@ -59,7 +59,7 @@ func AddWaitFlag(cmd *cobra.Command, wait *kube.WaitStrategy) {
 		newWaitValue(kube.HookOnlyStrategy, wait),
 		"wait",
-		"if specified, will wait until all resources are in the expected state before marking the operation as successful. It will wait for as long as --timeout. Valid inputs are 'watcher' and 'legacy'",
+		"wait until resources are ready (up to --timeout). Values: 'watcher' (default), 'legacy'. '--wait' without value = 'watcher'. Booleans deprecated; '--wait=false' does hook-only.",
```
I think your comment make it more relevant that `--wait` should also accept `hookOnly` which is not the case for the moment.

```
make build && ./bin/helm upgrade simple-test simple-test --debug --wait=hookOnly
go mod tidy
Error: invalid argument "hookOnly" for "--wait" flag: invalid wait input "hookOnly". Valid inputs are watcher, and legacy
```

I changed this in the last 2 commits.

benoittgt committed:
Avoid confusion between `--wait` (watcher) and no --wait (hookOnly)

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

TerryHowe reviewed (approved):
/lgtm

Not that I know too much about these strategies

benoittgt commented on pkg/kube/client.go:
```diff
@@ -88,12 +88,23 @@ type Client struct {
 
 const (
+	// StatusWatcherStrategy: event-driven waits using kstatus (watches + aggregated readers).
+	// Default for --wait. More accurate and responsive; waits CRs and full reconciliation.
```
Same I prefere without.

benoittgt commented on pkg/cmd/flags.go:
```diff
@@ -57,9 +57,9 @@ func addValueOptionsFlags(f *pflag.FlagSet, v *values.Options) {
 func AddWaitFlag(cmd *cobra.Command, wait *kube.WaitStrategy) {
 	cmd.Flags().Var(
-		newWaitValue(kube.HookOnlyStrategy, wait),
+		newWaitValue(kube.StatusWatcherStrategy, wait),
```
Isn't too late to change this for v4 ?

benoittgt committed:
Provide more help for SDK user when setting up WaitStrategy

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Do not change the default waiting strategy when --wait is not set

It is too much a breaking change for a new RC of beta V4.
See comment: https://github.com/helm/helm/pull/31421#discussion_r2465121438

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Prevent surprising failure with SDK when timeout is not set

While testing SDK features for v4. I was surprised with the error:

"reporter failed to start: event funnel closed: context deadline exceeded"

This occurs when no timeout is set:

```
upgradeClient := action.NewUpgrade(actionConfig)
upgradeClient.WaitStrategy = kube.StatusWatcherStrategy
// When Timeout is zero, the status wait uses a context with zero timeout which
// immediately expires causing "context deadline exceeded" errors.
upgradeClient.Timeout = 2 * time.Minute
```

With this patch it will work without specifying.

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Error strategy list match help

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

robertsirc reviewed (dismissed):
LGTM

gjenkins8 commented on pkg/cmd/flags.go:
```diff
@@ -59,7 +59,7 @@ func AddWaitFlag(cmd *cobra.Command, wait *kube.WaitStrategy) {
 		newWaitValue(kube.HookOnlyStrategy, wait),
 		"wait",
-		"if specified, will wait until all resources are in the expected state before marking the operation as successful. It will wait for as long as --timeout. Valid inputs are 'watcher' and 'legacy'",
+		"wait until resources are ready (up to --timeout). Values: 'watcher', 'hookOnly' (default), and 'legacy'",
```
As "Problem 2: Confusing CLI Mapping" points out, there are two "defaults". The default for when no flag is specified. And the default for when the flag is specified, but no value is provided (the result of convert from a boolean flag)

Thinking about how to represent these. Perhaps?:

```suggestion
		"if specified, wait until resources are ready (up to --timeout). Values: 'watcher' (default), 'hookOnly', and 'legacy'.",
```

benoittgt committed:
Update pkg/cmd/flags.go

Co-authored-by: George Jenkins <gvjenkins@gmail.com>
Signed-off-by: Benoit Tigeot <benoittgt@users.noreply.github.com>
Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

TerryHowe reviewed (approved):
/lgtm

scottrigby reviewed (approved):
Nice improvement. Also appreciate the help text improvement - much more clear about two defaults (for when flag isn't set and is set but with no value). Added `needs docs` label because I think we should ensure helm-www site notes or makes use of this somehow in the sdk examples / sdk docs section.
